src/Catalyst.Core.Modules.Keystore/Cms.cs

```diff
@@ -0,0 +1,173 @@
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation. The task's result is
+        ///   the plain text byte array of the protected data.
+        /// </returns>
+        /// <exception cref="KeyNotFoundException">
+        ///   When the required private key, to decrypt the data, is not foumd.
+        /// </exception>
+        /// <remarks>
+        ///   Cryptographic Message Syntax (CMS), aka PKCS #7 and 
+        ///   <see href="https://tools.ietf.org/html/rfc5652">RFC 5652</see>,
+        ///   describes an encapsulation syntax for data protection. It
+        ///   is used to digitally sign, digest, authenticate, and/or encrypt
+        ///   arbitrary message content.
+        /// </remarks>
+        public async Task<byte[]> ReadProtectedDataAsync(byte[] cipherText,
+            CancellationToken cancel = default(CancellationToken))
+        {
+            var cms = new CmsEnvelopedDataParser(cipherText);
+
+            // Find a recipient whose key we hold. We only deal with recipient names
+            // issued by ipfs (O=ipfs, OU=keystore).
+            var knownKeys = (await ListAsync(cancel).ConfigureAwait(false)).ToArray();
+            var recipient = cms.GetRecipientInfos().GetRecipients()
+               .OfType<RecipientInformation>()
+               .Select(ri =>
+                {
+                    var kid = GetKeyId(ri);
+                    var key = knownKeys.FirstOrDefault(k => k.Id == kid);
+                    return new
+                    {
+                        recipient = ri, key = key
+                    };
+                }).FirstOrDefault(r => r.key != null);
+
+            if (recipient == null)
+                throw new KeyNotFoundException("The required decryption key is missing.");
+
+            // Decrypt the contents.
+            var decryptionKey = await GetPrivateKeyAsync(recipient.key.Name, cancel).ConfigureAwait(false);
+            return recipient.recipient.GetContent(decryptionKey);
+        }
+
+        /// <summary>
+        ///   Get the key ID for a recipient.
+        /// </summary>
+        /// <param name="ri">
+        ///   A recepient of the message.
+        /// </param>
+        /// <returns>
+        ///   The key ID of the recepient or <b>null</b> if the recepient info
+        ///   is not understood or does not contain an IPFS key id.
+        /// </returns>
+        /// <remarks>
+        ///   The key ID is either the Subject Key Identifier (preferred) or the
+        ///   issuer's distinguished name with the form "CN=&lt;kid>,OU=keystore,O=ipfs".
+        /// </remarks>
+        MultiHash GetKeyId(RecipientInformation ri)
+        {
+            // Any errors are simply ignored.
+            try
+            {
+                // Subject Key Identifier is the key ID.
+                if (ri.RecipientID.SubjectKeyIdentifier is byte[] ski)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Framework/pullRequest?prid=4612924)